subsys/bluetooth/mesh/prov.c

```diff
@@ -131,89 +219,70 @@ int bt_mesh_prov_auth(uint8_t method, uint8_t action, uint8_t size)
 
 	case AUTH_METHOD_OUTPUT:
 		output = output_action(action);
-		if (!output) {
-			return -EINVAL;
-		}
 
-		if (!(bt_mesh_prov->output_actions & output)) {
-			return -EINVAL;
-		}
-
-		if (size > bt_mesh_prov->output_size) {
-			return -EINVAL;
+		if (!is_provisioner) {
+			err = check_output_auth(output, size);
+			if (err) {
+				return err;
+			}
 		}
 
-		atomic_set_bit(bt_mesh_prov_link.flags, NOTIFY_INPUT_COMPLETE);
-
-		if (output == BT_MESH_DISPLAY_STRING) {
-			unsigned char str[9];
-			uint8_t i;
-
-			bt_rand(str, size);
-
-			/* Normalize to '0' .. '9' & 'A' .. 'Z' */
-			for (i = 0U; i < size; i++) {
-				str[i] %= 36;
-				if (str[i] < 10) {
-					str[i] += '0';
-				} else {
-					str[i] += 'A' - 10;
-				}
-			}
-			str[size] = '\0';
-
-			memcpy(bt_mesh_prov_link.auth, str, size);
-			(void)memset(bt_mesh_prov_link.auth + size, 0,
-				     sizeof(bt_mesh_prov_link.auth) - size);
-
-			return bt_mesh_prov->output_string((char *)str);
-		} else {
-			uint32_t div[8] = { 10, 100, 1000, 10000, 100000,
-					    1000000, 10000000, 100000000 };
-			uint32_t num;
-
-			bt_rand(&num, sizeof(num));
-
-			if (output == BT_MESH_BLINK ||
-			    output == BT_MESH_BEEP ||
-			    output == BT_MESH_VIBRATE) {
-				/* According to the Bluetooth Mesh Profile
-				 * Specification Section 5.4.2.4, blink, beep
-				 * and vibrate should be a random integer
-				 * between 0 and 10^size, *exclusive*:
-				 */
-				num = (num % (div[size - 1] - 1)) + 1;
+		if (is_provisioner) {
+			if (output == BT_MESH_DISPLAY_STRING) {
+				input = BT_MESH_ENTER_STRING;
+				atomic_set_bit(bt_mesh_prov_link.flags, WAIT_STRING);
 			} else {
-				num %= div[size - 1];
+				input = BT_MESH_ENTER_NUMBER;
+				atomic_set_bit(bt_mesh_prov_link.flags, WAIT_NUMBER);
 			}
 
-			sys_put_be32(num, &bt_mesh_prov_link.auth[12]);
-			(void)memset(bt_mesh_prov_link.auth, 0, 12);
+			return bt_mesh_prov->input(input, size);
+		}
 
-			return bt_mesh_prov->output_number(output, num);
+		if (output == BT_MESH_DISPLAY_STRING) {
+			char str[9];
+
+			atomic_set_bit(bt_mesh_prov_link.flags, NOTIFY_INPUT_COMPLETE);
+			get_auth_string(str, size);
+			return bt_mesh_prov->output_string(str);
 		}
 
+		atomic_set_bit(bt_mesh_prov_link.flags, NOTIFY_INPUT_COMPLETE);
+		return bt_mesh_prov->output_number(output,
+				get_auth_number(output, BT_MESH_NO_INPUT, size));
+
 	case AUTH_METHOD_INPUT:
 		input = input_action(action);
-		if (!input) {
-			return -EINVAL;
-		}
 
-		if (!(bt_mesh_prov->input_actions & input)) {
-			return -EINVAL;
+		if (!is_provisioner) {
+			err = check_input_auth(input, size);
+			if (err) {
+				return err;
+			}
 		}
 
-		if (size > bt_mesh_prov->input_size) {
-			return -EINVAL;
+		if (!is_provisioner) {
```

Comment:
Combine into one `if`

Reply:
looks like you requested to do this change 5 days ago https://github.com/zephyrproject-rtos/zephyr/pull/38205#discussion_r700189030
Now you ask to combine back. Didn't I understand anything?

Reply:
Never mind, got it. will fix